File: 4_leetcode/C++/LC_1195_sem.cpp
#include <semaphore>
#include <iostream>
#include <thread>

#define NUM_THREADS 4

class FizzBuzz {
private:
    int n;
    int tick;
    bool end;
    std::binary_semaphore sem_mutex;

public:
    FizzBuzz(int n) : 
        n(n), 
        tick(1), 
        end(false), 
        sem_mutex(1) {}
    

    void fizz(function<void()> printFizz) {
        while(true) {
            sem_mutex.acquire();

            while((tick%3!=0 || tick%5==0) && !end) {
                sem_mutex.release();
                sem_mutex.acquire();
            }

            if (tick>n || end) {
                end=true;
                for(int i=0;i<NUM_THREADS;i++) sem_mutex.release();

                break;
            }

            printFizz();
            tick++;
            for(int i=0;i<NUM_THREADS;i++) sem_mutex.release();
        }
    }

        
    void buzz(function<void()> printBuzz) {
        while(true) {
            sem_mutex.acquire();

            while((tick%5!=0 || tick%3==0) && !end) {
                sem_mutex.release();
                sem_mutex.acquire();
            }

            if (tick>n || end) {
                end=true;
                for(int i=0;i<NUM_THREADS;i++) sem_mutex.release();

                break;
            }

            printBuzz();
            tick++;
            for(int i=0;i<NUM_THREADS;i++) sem_mutex.release();
        }
    }
    
    void fizzbuzz(function<void()> printFizzBuzz) {
        while(true) {
            sem_mutex.acquire();

            while((tick%3!=0 || tick%5!=0) && !end) {
                sem_mutex.release();
                sem_mutex.acquire();
            }

            if (tick>n || end) {
                end=true;
                for(int i=0;i<NUM_THREADS;i++) sem_mutex.release();

                break;
            }

            printFizzBuzz();
            tick++;            
            for(int i=0;i<NUM_THREADS;i++) sem_mutex.release();
        }
    }

    void number(function<void(int)> printNumber) {
        while(true) {
            sem_mutex.acquire();

            while((tick%3==0 || tick%5==0) && !end) {
                sem_mutex.release();
                sem_mutex.acquire();
            }

            if (tick>n || end) {
                end=true;
                for(int i=0;i<NUM_THREADS;i++) sem_mutex.release();

                break;
            }

            printNumber(tick);
            tick++;
            for(int i=0;i<NUM_THREADS;i++) sem_mutex.release();
        }
    }
};


        

File: 4_leetcode/C#/LC_1195_sem.cs
using System;
using System.Threading;

public class FizzBuzz {
    private int n;

    private const int NUM_THREADS=4;
    private int tick;
    private bool end;
    private SemaphoreSlim semMutex;

    public FizzBuzz(int n) {      
        this.n =n;
        tick   =1;
        end    =false;

        semMutex=new SemaphoreSlim(1, 1);
    }

    // printFizz() outputs "fizz".
    public void Fizz(Action printFizz) {
        while (true) {
            semMutex.Wait();

            try {
                while ((tick%3!=0 || tick%5==0) && !end) {
                    semMutex.Release();
                    semMutex.Wait();
                }

                if (tick>n || end) {
                    end=true;
                    break;
                }

                printFizz();
                tick++;
            }
            finally {
                semMutex.Release();
            }
        }
    }

    // printBuzzz() outputs "buzz".
    public void Buzz(Action printBuzz) {
        
        while (true) {
            semMutex.Wait();

            try {
                while ((tick%5!=0 || tick%3==0) && !end) {
                    semMutex.Release();
                    semMutex.Wait();
                }

                if (tick>n || end) {
                    end=true;
                    break;
                }

                printBuzz();
                tick++;
            }
            finally {
                semMutex.Release();
            }
        }
    }

    // printFizzBuzz() outputs "fizzbuzz".
    public void Fizzbuzz(Action printFizzBuzz) {
        while (true) {
            semMutex.Wait();

            try {
                while ((tick%3!=0 || tick%5!=0) && !end) {
                    semMutex.Release();
                    semMutex.Wait();
                }

                if (tick>n || end) {
                    end=true;
                    break;
                }

                printFizzBuzz();
                tick++;
            }
            finally {
                semMutex.Release();
            }
        }
    }

    // printNumber(x) outputs "x", where x is an integer.
    public void Number(Action<int> printNumber) {
        while (true) {
            semMutex.Wait();

            try {
                while ((tick%3==0 || tick%5==0) && !end) {
                    semMutex.Release();
                    semMutex.Wait();
                }

                if (tick>n || end) {
                    end=true;
                    break;
                }

                printNumber(tick);
                tick++;
            }
            finally {
                semMutex.Release();
            }
        }
    }
}